File: algorithms/3_hard/P/02551_put_marbles_in_bags/solution.c
static int cmp(const void *a_ptr, const void *b_ptr) {
    const int a = *(const int*)a_ptr;
    const int b = *(const int*)b_ptr;
    return (a > b) - (a < b);
}

long long putMarbles(const int *restrict weights, const int weights_len, const int k) {
    if (k >= weights_len || k == 1) { return 0; }

    const int scores_len = weights_len - 1; 
    int *restrict scores = malloc((size_t)scores_len * sizeof(*scores));
    if (scores == NULL) { return -1; }

    for (int i = 0; i < weights_len - 1; i += 1) {
        scores[i] = weights[i] + weights[i + 1];
    }

    qsort(scores, (size_t)scores_len, sizeof(*scores), cmp);

    int64_t max = 0;
    int64_t min = 0;

    for (int i = 0; i < (k - 1); i += 1) {
        max += scores[scores_len - i - 1];
        min += scores[i];
    }

    free(scores);
    return max - min;
}

File: algorithms/3_hard/P/02551_put_marbles_in_bags/solution.rs
impl Solution {
    pub fn put_marbles(weights: Vec<i32>, k: i32) -> i64 {
        let k = k as usize;
        if k >= weights.len() || k == 1 { return 0; }

        let mut scores = Vec::with_capacity(weights.len() - 1);

        for w in weights.windows(2) {
            scores.push(w[0] + w[1]);
        }

        scores.sort_unstable();

        let scores_len = scores.len();
        let max: i64 = scores[(scores_len - (k - 1))..scores_len].iter().map(|&x| x as i64).sum();
        let min: i64 = scores[0..(k - 1)].iter().map(|&x| x as i64).sum();

        max - min
    }
}
